### PR TITLE
Add clipping to rotor_vel

### PR DESCRIPTION
**Pull request by ratheron**
Before there was no explicit clipping of the rotor speed/thrust state. This PR adds this by
1. implementing clipping on the velocity state **after** integration, such that an overshoot in the integration state does not exceed the limits
2. implementing clipping **inside** the integration step, such that the derivative is clipped in case we are already at the bounds to prevent further overshoots.

There are two minor problems with this approach
1. the clipping inside the integration step cannot simply be removed from the step pipeline, since it's baked in.
2. in case the state is still in bounds, a large command is applied, and an integration method besides Euler is used, the effective acceleration on the mass can still exceed the actuator limits for one sim step.

### CODE REVIEW

File: crazyflow/sim/integration.py [hunk 1]
```diff
@@ -1,3 +1,15 @@
+"""Numerical integrators for the simulation dynamics.
+
+All integrators advance the full drone state, including the rotor state ``rotor_vel``, by one
+simulation step.
+
```
Comment (amacati) [inline crazyflow/sim/integration.py]:
```suggestion
```

File: crazyflow/sim/integration.py [hunk 2]
```diff
@@ -1,3 +1,15 @@
...
+Note:
+    The rotor limits are only enforced by clipping ``rotor_vel`` after the integration (see
+    [clip_rotor_vel][crazyflow.sim.sim.clip_rotor_vel]). Higher order integration methods like RK4
+    evaluate the dynamics at intermediate states that are not clipped, so the ``rotor_vel`` seen by
+    the model can transiently exceed the limits within a single step.
```
Comment (amacati) [inline crazyflow/sim/integration.py]:
```suggestion
    State limits are enforced by clipping after the integration (see e.g.
    [clip_rotor_vel][crazyflow.sim.sim.clip_rotor_vel]). The derivatives are unclipped. Higher order
    integration methods like RK4 evaluate the dynamics at intermediate states, so derivatives seen by
    the model can transiently exceed the limits within a single step.
```

File: docs/user-guide/pipelines.md [hunk 3]
```diff
@@ -130,6 +131,9 @@ remove_fn(sim.step_pipeline, "clip_floor_pos")
 sim.build_step_fn()
 ```
 
+!!! note
+    The rotor limits are enforced solely by the `clip_rotor_vel` stage, which clips the rotor state after each integration step. Higher order integration methods like RK4 evaluate the dynamics at intermediate, unclipped states, so the `rotor_vel` seen by the model can transiently exceed the limits within a single step.
+
```
Comment (amacati) [inline docs/user-guide/pipelines.md:136]:
```suggestion
!!! note
    The rotor limits are enforced by the `clip_rotor_vel` stage, which clips the rotor state after each integration step. Higher order integration methods like RK4 evaluate the dynamics at intermediate, unclipped states, so the `rotor_vel` seen by the model can transiently exceed the limits within a single step.

```

File: tests/unit/test_sim.py [hunk 4]
```diff
@@ -396,6 +397,62 @@ def test_floor_penetration(dynamics: Dynamics):
...
+@pytest.mark.parametrize(
+    "dynamics", [Dynamics.first_principles, Dynamics.so_rpy_rotor, Dynamics.so_rpy_rotor_drag]
+)
+@pytest.mark.parametrize("integrator", Integrator)
+def test_rotor_vel_clip(dynamics: Dynamics, integrator: Integrator):
+    """Test that the rotor state (RPM or thrust) saturates at its physical limits.
+
+    Commands beyond the limits must not push the state past them, not even transiently, and states
+    that already violate the limits (e.g. set by the user) must be clipped after one step.
+    """
```
Comment (amacati) [inline tests/unit/test_sim.py]:
```suggestion
    """Test that the rotor state (RPM or thrust) saturates at its physical limits."""
```

File: tests/unit/test_sim.py [hunk 5]
```diff
@@ -396,6 +397,62 @@ def test_floor_penetration(dynamics: Dynamics):
...
+    control = Control.rotor_vel if dynamics == Dynamics.first_principles else Control.attitude
+    sim = Sim(dynamics=dynamics, control=control, integrator=integrator, device="cpu")
+    lower, upper = rotor_vel_limits(dynamics, sim.drone)
+    assert 0.0 < lower < upper
+    stages = tuple(sim.step_pipeline)
+    assert stages.index("clip_rotor_vel") == stages.index("integration") + 1
```
Comment (amacati) [inline tests/unit/test_sim.py]: This is not necessary. We might add more clippings in the future, and the index might shift
```suggestion
```

File: tests/unit/test_sim.py [hunk 6]
```diff
@@ -396,6 +397,62 @@ def test_floor_penetration(dynamics: Dynamics):
...
+        assert jnp.all(rotor_vel() == lower), f"rotor_vel {rotor_vel()} did not saturate at {lower}"
+    # States outside the limits are clipped after the integration
+    for value, target in ((2 * upper, upper), (-upper, lower)):
+        command(target)
+        states = sim.data.states.replace(rotor_vel=jnp.full_like(sim.data.states.rotor_vel, value))
+        sim.data = sim.data.replace(states=states)
+        sim.step()
+        assert jnp.all(sim.data.states.rotor_vel >= lower)
+        assert jnp.all(sim.data.states.rotor_vel <= upper)
+    sim.close()
```
Comment (amacati) [inline tests/unit/test_sim.py:442]: Too many special cases imo. Can you split this in `test_rotor_vel_clip` and `test_thrust_clip`? The first only uses first principles, and the other only the `so_rpy` models. Btw, why not test this over all `so_rpy` models?

Splitting also gets rid of the definitions like `command` and `rotor_vel`.

File: examples/jax/gradient_clipping.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+def plot_results(
+    cmds: NDArray,
+    results: dict[str, tuple[NDArray, NDArray]],
+    lower: float,
+    upper: float,
+    freq: int,
+):
+    # Only import if plotting is desired to avoid a dependency on matplotlib
+    import matplotlib.pyplot as plt
```
Comment (amacati) [inline examples/jax/gradient_clipping.py:88]: This should go on top. We have matplotlib in the test envs and import it at the top of the file in all other tests

File: examples/jax/gradient_clipping.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    # Option 1: keep the clipping as is. The rotor state respects the limits, but the gradient is
+    # zero while the state is saturated
+    results = {"clip (default)": rollout(sim, cmds)}
+
+    # Option 2: clip the state in the forward pass, but keep the gradients flowing in the backward
+    # pass (straight-through estimator)
+    clip_fn = partial(clip_rotor_vel_nonblocking, lower=lower, upper=upper)
+    replace_fn(sim.step_pipeline, clip_fn, "clip_rotor_vel")
+    results["nonblocking clip"] = rollout(sim, cmds)
+
```
Comment (amacati) [inline examples/jax/gradient_clipping.py:70]: Image

**Review by amacati — CHANGES_REQUESTED** [review]

File: examples/jax/gradient_clipping.py (hunk 7 above)
Comment (ratheron) [reply examples/jax/gradient_clipping.py]: But people may run the tests with the base env, which is why it is only imported when needed in all examples.

Comment (ratheron) [reply examples/jax/gradient_clipping.py:88]: But people may run the tests with the base env, which is why it is only imported when needed in all examples.

**Review by amacati — APPROVED** [review]
